### ISSUE TITLE
Inefficient response handling in Snap

### ISSUE THREAD

**Issue by holiman** [ethereum/go-ethereum#25600]
**Note** This bug report is somewhat speculative, I am not 100% certain that the description here is correct. 

------------------------

I think I know why this PR (https://github.com/ethereum/go-ethereum/pull/25588) does not make any difference. 
![Screenshot 2022-08-25 at 08-44-34 Single Geth - Grafana](https://user-images.githubusercontent.com/142290/186594071-4fbbfb52-4272-4471-abe7-8c0c79e8c0ec.png)
This is the machine charts. A few observations can be made: 

- There is always 128-150 pending (outstanding) requests
- The `Snap trienode handle` times are most commonly in `3-7s` bucket, but quite often up to 20 seconds. 

So if all our responses have already arrived, handling them will take ~`1280` seconds, or 21 minutes. So once we get to actually handle them, they will have timed out in the snap layer, although they were fine in the p2p packet tracker layer. 

This is wasteful: we are making requests at a higher rate than we can handle, and thus ignoring responses and refetching stuff. We need to adjust the mechanism so that we do not request more than we can handle -- alternatively, fix the timeout management so that we do not time out deliveries which have already arrived and are just waiting in the queue.

**Comment by holiman** [comment]
In `sync.go,` every time the loop runs, we will assign/issue another trienode heal request. So once an item has ben lying on the queue for ~15 minutes, we issue send out request. It may be served quickly, but will not be handled until after it has timed out. 

A better model in this case would be to tune down the `128` pending ones down to maybe `10`. 

OBS: this is a consequence of disk IO speed, not network lag.

**Comment by holiman** [comment]
Here's one hour of execution, where I have set a hard upper limit of pending trie requests at `15`. 
![Screenshot 2022-08-25 at 14-01-59 Single Geth - Grafana](https://user-images.githubusercontent.com/142290/186659233-c1e47db6-ba8c-4a50-afaf-9301c79fb77b.png)

The `Unexpected trienode heal` log messages almost (but not quite) go away completely.
![Screenshot 2022-08-25 at 14-05-57 bootnode-azure-koreasouth-001 - Papertrail](https://user-images.githubusercontent.com/142290/186659936-8c86505a-e425-40c1-96a7-e23e70938eac.png)
The code I used is at this branch: https://github.com/holiman/go-ethereum/tree/heal_timeout_fix . 

It places the cap very un-dynamically at the place where we schedule requests. It could be applied in several different places, and could also be made dynamic -- e.g. adjusted so that the `maxPending` goes down if the mean handle-time goes up. Suggestions appreciated.
